cluster-autoscaler/cloudprovider/aws/managed_nodegroup_cache.go

```diff
@@ -0,0 +1,153 @@
+// This allows to get a better repartition of the AWS queries.
+type managedNodegroupCache struct {
+	cache.Store
+	mngJitterClock clock.Clock
+	awsService     *awsWrapper
+}
+
+// This struct will be used to hold some information from the describeNodegroup call
+// There are more options that can be added in the future
+// https://docs.aws.amazon.com/cli/latest/reference/eks/describe-nodegroup.html
+type managedNodegroupCachedObject struct {
+	name        string
+	clusterName string
+	taints      []apiv1.Taint
+	labels      map[string]string
+}
+
+type mngJitterClock struct {
+	clock.Clock
+
+	jitter bool
+	sync.RWMutex
+}
+
+func newManagedNodeGroupCache(awsService *awsWrapper) *managedNodegroupCache {
+	jc := &mngJitterClock{}
+	return newManagedNodeGroupCacheWithClock(
+		awsService,
+		jc,
+		cache.NewExpirationStore(func(obj interface{}) (s string, e error) {
+			return obj.(managedNodegroupCachedObject).name, nil
+		}, &cache.TTLPolicy{
+			TTL:   managedNodegroupCachedTTL,
+			Clock: jc,
+		}),
+	)
+}
+
+func newManagedNodeGroupCacheWithClock(awsService *awsWrapper, jc clock.Clock, store cache.Store) *managedNodegroupCache {
+	return &managedNodegroupCache{
+		store,
+		jc,
+		awsService,
+	}
+}
+
+func (c *mngJitterClock) Since(ts time.Time) time.Duration {
+	since := time.Since(ts)
+	c.RLock()
+	defer c.RUnlock()
+	if c.jitter {
+		return since + (time.Second * time.Duration(rand.IntnRange(mngCacheMinTTL, mngCacheMaxTTL)))
+	}
+	return since
+}
+
+func (m *managedNodegroupCache) getManagedNodegroup(nodegroupName string, clusterName string) (*managedNodegroupCachedObject, error) {
+	taintList, labelMap, err := m.awsService.getManagedNodegroupInfo(nodegroupName, clusterName)
+	if err != nil {
+		// If there's an error cache an empty nodegroup to limit failed calls to the EKS API
+		newEmptyNodegroup := managedNodegroupCachedObject{
+			name:        nodegroupName,
+			clusterName: clusterName,
+			taints:      nil,
+			labels:      nil,
+		}
+
+		m.Add(newEmptyNodegroup)
+		return nil, err
+	}
+
+	newNodegroup := managedNodegroupCachedObject{
+		name:        nodegroupName,
+		clusterName: clusterName,
+		taints:      taintList,
+		labels:      labelMap,
+	}
+
+	m.Add(newNodegroup)
+
+	return &newNodegroup, nil
+}
+
+func (m managedNodegroupCache) getManagedNodegroupLabels(nodegroupName string, clusterName string) (map[string]string, error) {
+	// List expires old entries
+	cacheList := m.List()
+	klog.V(5).Infof("Current ManagedNodegroup cache: %+v\n", cacheList)
+
+	if obj, found, _ := m.GetByKey(nodegroupName); found {
+		return obj.(managedNodegroupCachedObject).labels, nil
+	}
+
+	managedNodegroupInfo, err := m.getManagedNodegroup(nodegroupName, clusterName)
+	if err != nil {
+		klog.Errorf("Failed to query the managed nodegroup %s for the cluster %s while looking for labels: %v", nodegroupName, clusterName, err)
+		return nil, err
+	}
+	return managedNodegroupInfo.labels, nil
```

Comment:
Much of the logic in this and the following function is replicated, can we condense the majority of this down to reduce repetition.

Reply:
I can look into it this.